src/Constraint/Contains.php

```diff
@@ -64,7 +64,7 @@ public function evaluate($value)
     public function equivalentTo(Expression $other)
     {
         // Since this class is final, we can check with instanceof
-        return $other instanceof $this && $this->comparedValue == $other->comparedValue;
+        return $other instanceof $this && (string) $this->comparedValue === (string) $other->comparedValue;
```

Comment:
Well, your constructor documents that it expects a string. If you actually accept other kind of values without considered it is a user mistake, then the current implementation was indeed buggy, and the casting to string is needed